lib/using_restricted_array.rb

```diff
@@ -1,61 +1,121 @@
 # RestrictedArray can be created using a specified size, or a random size in
 # the range of 1-20 will be chosen for you.
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: ? O(n) Because it loops through based on the input size (array size)
+# Space complexity: ? O(1)
 def length(array)
-  raise NotImplementedError
+  size = 0
+
+  while array[size] != nil
+    size += 1
+  end
+  return size
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n)
+# Space complexity: ? O(1)
 def print_array(array)
-  raise NotImplementedError
+  i = 0
+
+  while array[i] != nil
+    print array[i]
+    print " "
+    i += 1
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n)
+# Space complexity: ? O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  return false if length == 0
+
+  i = 0
+  while i < length
+    return true if array[i] == value_to_find
+    i += 1
+  end
+
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n)
+# Space complexity: ? O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  return nil if length == 0
+  i = 1
+  max = array[0]
+  while i < length(array)
+    max = array[i] if array[i] > max
+    i += 1
+  end
+  return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n)
+# Space complexity: ? O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  return nil if length == 0
+  min = array[0]
+  i = 1
+
+  while i < length(array)
+    min = array[i] if array[i] < min
+    i += 1
+  end
+  return min
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n)
+# Space complexity: ? O(1) modifying the original array
 def reverse(array, length)
-  raise NotImplementedError
+  first_index = 0
+  last_index = length - 1
+  while first_index < last_index
+    temp = array[first_index]
+    array[first_index] = array[last_index]
+    array[last_index] = temp
+    first_index += 1
+    last_index -= 1
+  end
+  return
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(log n)
+# Space complexity: ? O(1)
 def binary_search(array, length, value_to_find)
```

Comment:
This is close, but it's not binary search.  Consider for an array of 5 elements [0, 1, 2, 3, 4, 5, 6, 7, 8, 9], if we're searching for 5.  

Midpoint starts at index 5 (value 5), which is too small so midpoint becomes 8 and that value is too large so doing the math it becomes 4 which is smaller than a value previously already checked.  

Consider

```ruby
high = length - 1
low = 0
while low <= high
  midpoint = (low + high) / 2
  
  return true if array[midpoint] == value_to_find)

   if array[midpoint] > value_to_find
      high = midpoint - 1
    else
      low = midpoint + 1
    end
end

return false